vkcs/compute/data_source_flavor.go

```diff
@@ -137,100 +232,121 @@ func dataSourceComputeFlavorRead(ctx context.Context, d *schema.ResourceData, me
 		}
 
-		allFlavors = append(allFlavors, *flavor)
-	} else {
-		accessType := flavors.AllAccess
-		if v, ok := d.GetOk("is_public"); ok {
-			if v, ok := v.(bool); ok {
-				if v {
-					accessType = flavors.PublicAccess
-				} else {
-					accessType = flavors.PrivateAccess
-				}
-			}
-		}
-		listOpts := flavors.ListOpts{
-			MinDisk:    d.Get("min_disk").(int),
-			MinRAM:     d.Get("min_ram").(int),
-			AccessType: accessType,
-		}
+		return diag.FromErr(dataSourceComputeFlavorAttributes(d, computeClient, flavor, nil))
+	}
 
-		log.Printf("[DEBUG] vkcs_compute_flavor ListOpts: %#v", listOpts)
+	requiredFlavor := NewRequiredFlavorFromResourceData(d)
+	listOpts := flavors.ListOpts{
+		MinDisk:    requiredFlavor.MinDisk,
+		MinRAM:     requiredFlavor.MinRAM,
+		AccessType: requiredFlavor.AccessType,
+	}
 
-		allPages, err := flavors.ListDetail(computeClient, listOpts).AllPages()
-		if err != nil {
-			return diag.Errorf("Unable to query VKCS flavors: %s", err)
-		}
+	log.Printf("[DEBUG] vkcs_compute_flavor ListOpts: %#v", listOpts)
 
-		allFlavors, err = flavors.ExtractFlavors(allPages)
-		if err != nil {
-			return diag.Errorf("Unable to retrieve VKCS flavors: %s", err)
-		}
+	allPages, err := flavors.ListDetail(computeClient, listOpts).AllPages()
+	if err != nil {
+		return diag.Errorf("Unable to query VKCS flavors: %s", err)
+	}
+
+	allFlavors, err := iflavors.ExtractFlavorWithExtraSpecs(allPages)
+	if err != nil {
+		return diag.Errorf("Unable to retrieve VKCS flavors: %s", err)
 	}
 
 	// Loop through all flavors to find a more specific one.
 	if len(allFlavors) > 0 {
-		var filteredFlavors []flavors.Flavor
+		var filteredFlavors []iflavors.FlavorWithExtraFields
 		for _, flavor := range allFlavors {
-			if v := d.Get("name").(string); v != "" {
-				if flavor.Name != v {
-					continue
-				}
+			switch {
+			case requiredFlavor.HasName && flavor.Name != requiredFlavor.Name:
+				continue
+			case requiredFlavor.HasRAM && flavor.RAM != requiredFlavor.RAM:
+				continue
+			case requiredFlavor.HasVCPUs && flavor.VCPUs != requiredFlavor.VCPUs:
+				continue
+			case requiredFlavor.HasDisk && flavor.Disk != requiredFlavor.Disk:
+				continue
+			case requiredFlavor.HasSwap && flavor.Swap != requiredFlavor.Swap:
+				continue
+			case requiredFlavor.HasRxTxFactor && flavor.RxTxFactor != requiredFlavor.RxTxFactor:
+				continue
+			case requiredFlavor.HasExtraSpecs && flavor.FlavorExtraFields.ExtraSpecs == nil:
+				continue
 			}
-
-			// d.GetOk is used because 0 might be a valid choice.
-			if v, ok := d.GetOk("ram"); ok {
-				if flavor.RAM != v.(int) {
-					continue
-				}
+			if !requiredFlavor.HasExtraSpecs {
+				filteredFlavors = append(filteredFlavors, flavor)
+				continue
 			}
 
-			if v, ok := d.GetOk("vcpus"); ok {
-				if flavor.VCPUs != v.(int) {
-					continue
+			match := true
+			for spec, reqVal := range requiredFlavor.ExtraSpecs {
+				val, ok := flavor.ExtraSpecs[spec]
+				if !ok || !reflect.DeepEqual(val, reqVal) {
```

Comment:
Are there any non-string values in this map? It appears that Compute API can only produce map[string]string.

Reply:
Yes, @ftersin show me case with map in map, additionally this works for all future changes

Reply:
Show me, please, I've just checked all public flavors, and there are only extra specs with values of type string. They are also stored this way, and I don't see any way to get values of other types, you can also check OS doc: https://docs.openstack.org/api-ref/compute/#flavors-extra-specs-flavors-os-flavor-extra-specs. It's better to not try to cover all possible future scenarios, especially those that are unlikely to be implemented. I don't worry much about speed of reflection in this case, but it's generally better to avoid using it when you can do without it.

Reply:
Initially, the idea was to make this field in such a way that we did not have to change it with changes in openstack, and users worked directly with the openstack parameters and their documentation